lightning/src/routing/router.rs

```diff
@@ -414,25 +459,50 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, paye
+		HashMap::with_capacity(if first_hops.is_some() { first_hops.as_ref().unwrap().len() } else { 0 });
 	if let Some(hops) = first_hops {
 		for chan in hops {
 			let short_channel_id = chan.short_channel_id.expect("first_hops should be filled in with usable channels, not pending ones");
 			if chan.remote_network_id == *our_node_id {
 				return Err(LightningError{err: "First hop cannot have our_node_id as a destination.".to_owned(), action: ErrorAction::IgnoreError});
 			}
-			first_hop_targets.insert(chan.remote_network_id, (short_channel_id, chan.counterparty_features.clone(), chan.outbound_capacity_msat));
+			first_hop_targets.insert(chan.remote_network_id, (short_channel_id, chan.counterparty_features.to_context(), chan.outbound_capacity_msat, chan.counterparty_features.to_context()));
 		}
 		if first_hop_targets.is_empty() {
 			return Err(LightningError{err: "Cannot route when there are no outbound routes away from us".to_owned(), action: ErrorAction::IgnoreError});
 		}
 	}
 
+	let empty_channel_features = ChannelFeatures::empty();
+
+	// The main heap containing all candidate next-hops sorted by their score (max(A* fee,
+	// htlc_minimum)). Ideally this would be a heap which allowed cheap score reduction instead of
+	// adding duplicate entries when we find a better path to a given node.
+	let mut targets = BinaryHeap::new();
+
+	// Map from node_id to information about the best current path to that node, including feerate
+	// information.
+	let mut dist = HashMap::with_capacity(network.get_nodes().len());
+
+	// During routing, if we ignore a path due to an htlc_minimum_msat limit, we set this,
+	// indicating that we may wish to try again with a higher value, potentially paying to meet an
+	// htlc_minimum with extra fees while still finding a cheaper path.
+	let mut hit_minimum_limit;
+
+	// When arranging a route, we select multiple paths so that we can make a multi-path payment.
+	// We start with a path_value of the exact amount we want, and if that generates a route we may
+	// return it immediately. Otherwise, we don't stop searching for paths until we have 3x the
+	// amount we want in total across paths, selecting the best subset at the end.
+	const ROUTE_CAPACITY_PROVISION_FACTOR: u64 = 3;
+	let recommended_value_msat = final_value_msat * ROUTE_CAPACITY_PROVISION_FACTOR as u64;
+	let mut path_value_msat = final_value_msat;
+
 	// We don't want multiple paths (as per MPP) share liquidity of the same channels.
 	// This map allows paths to be aware of the channel use by other paths in the same call.
 	// This would help to make a better path finding decisions and not "overbook" channels.
 	// It is unaware of the directions (except for `outbound_capacity_msat` in `first_hops`).
-	let mut bookkeeped_channels_liquidity_available_msat = HashMap::new();
+	let mut bookkeeped_channels_liquidity_available_msat = HashMap::with_capacity(network.get_nodes().len());
```

Comment:
Pre-allocating to the size of the whole graph assumes we'll completely traverse it before reaching `recommended_value_msat` or `path_value_msat` liquidity threshold ? Or what's your thinking there ?

Reply:
Yea, I mean I guess we could divide by some factor, but why bother, its short-lived and at worst we will, so we might as well just allocate for it.